### PR TITLE
Simplify _png extension by handling file open/close in Python.

### PR DESCRIPTION
**Pull request by anntzer**
Make functions in the (private) _png extension only take file-like
objects as arguments, rather than "filename or file-like".  Filenames
were anyways handled by calling the python C-API to open/close the
files, so we may as well just do this from Python where a simple
`with open(...):` does the job.  The public API is not changed.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backends/backend_pgf.py [hunk 1]
```diff
@@ -639,7 +639,8 @@ def draw_image(self, gc, x, y, im, transform=None):
         fname = os.path.splitext(os.path.basename(self.fh.name))[0]
         fname_img = "%s-img%d.png" % (fname, self.image_counter)
         self.image_counter += 1
-        _png.write_png(im[::-1], os.path.join(path, fname_img))
+        with pathlib.Path(path, fname_img).open("wb") as file:
```
Comment (timhoffm) [inline lib/matplotlib/backends/backend_pgf.py]: I don't quite like assigning to builtin names, but acknowledge that there's already precedence for this in the library.

Comment (anntzer) [reply lib/matplotlib/backends/backend_pgf.py]: It's not a builtin name anymore...

**Comment by moorepants** [comment]
> The public API is not changed.

Even though the public API hasn't changed, i think this does seem to change the error types raised and has broken downstream code in SymPy. See this issue: https://github.com/sympy/sympy/issues/18815

The offending changed lines are:

https://github.com/matplotlib/matplotlib/blame/v3.2.0/lib/matplotlib/mathtext.py#L3436

I haven't determined yet if it is ultimately SymPy's bad design for relying on a specific error but this PR might cause all Jupyter notebook printing for SymPy to break once people install matplotlib 3.2.0.

**Comment by tacaswell** [comment]
Can you try with the 3.2.x branch?  We just merged https://github.com/matplotlib/matplotlib/pull/16709 which I think is related.

**Comment by QuLogic** [comment]
Yes, this is the same bug; see backtrace on linked issue.

**Pull request merged**